### PR TITLE
perf: introduce high-performance custom allocator replacing malloc/calloc/free

### PR DESCRIPTION
**Pull request by ShashwatAgrawal20**
Add cml_allocator (thread-cached size-class segregated fit allocator with slab backing and direct large-object path via posix_memalign/malloc).

Replace virtually all direct libc heap calls across the entire repo (src/, examples/, tests/, benchmarks/, main.c) with the new cml_malloc, cml_calloc, cml_realloc, cml_free and cml_strdup (plus aligned variants).

- New files: include/alloc/cml_allocator.h, src/alloc/cml_allocator.c
- Update CMakeLists.txt to build the new allocator
- Pull allocator declarations in via include/cml.h and core/logging.h (robust against platform #ifdef include blocks in various .c files)
- Legacy cml_safe_* wrappers now delegate to the fast path
- Minor supporting fixes (include ordering, AVX target attributes for certain SIMD paths to keep builds happy)

Goal: much faster allocation paths with lower variance for ML use cases (tensors, IR graphs, layers, tokenizers, etc.).

The allocator itself still uses raw malloc/mmap only for its internal slab acquisition (standard and required).

### CODE REVIEW

**Comment by jaywyawhare** [comment]
Thanks for the contribution boii.

**Comment by jaywyawhare** [comment]
@coderabbitai

**Comment by jaywyawhare** [comment]
@coderabbitai review the changes

File: src/alloc/cml_allocator.c [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+    size_t alloc_sz = sizeof(Slab) + num_blocks * (CML_HEADER_SIZE + bin_size);
+    /* Align the slab allocation a bit */
+    void* raw = system_malloc(alloc_sz + 64);
+    if (!raw) return NULL;
+
+    /* Align start of Slab data area for cleanliness */
+    uintptr_t base = (uintptr_t)raw;
+    uintptr_t aligned_base = (base + 63) & ~(uintptr_t)63;
+    Slab* slab = (Slab*)aligned_base;
+
```
Comment (jaywyawhare) [inline src/alloc/cml_allocator.c:216]: raw pointer from system_malloc L198 is lost after computing the aligned offset L202-204, only the aligned slab pointer is stored and returned, making it impossible to later call system_free(raw)

File: src/alloc/cml_allocator.c [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+#define CML_HEADER_SIZE        16             /* 16B header => good default alignment for returned ptrs */
+
+_Static_assert(CML_HEADER_SIZE >= 16 && (CML_HEADER_SIZE % 16) == 0, "header must preserve alignment");
+
+/* Per-allocation header. Lives immediately before user pointer. */
+typedef struct {
+    uint32_t size;       /* requested user bytes */
+    uint16_t class_idx;  /* which size class (0xffff for large/direct) */
+    uint16_t magic;      /* 0xC4A1 for sanity */
+} AllocHeader;
```
Comment (jaywyawhare) [inline src/alloc/cml_allocator.c]: uint32_t limiting individual allocations around 4GB ML workloads commonly require larger tensors try using size_t

File: src/nn/layers/rnn.c [hunk 3]
```diff
@@ -816,14 +817,14 @@ void gru_forward(GRU* gru, Tensor* input, Tensor* h_0,
...
-    Tensor** final_h = calloc((size_t)total_dirs, sizeof(Tensor*));
+    Tensor** final_h = cml_calloc((size_t)total_dirs, sizeof(Tensor*));
     Tensor* layer_input = x;
 
     for (int l = 0; l < gru->num_layers; l++) {
         GRUCell* fwd_cell = gru->cells[l * nd + 0];
 
         Tensor* h_fwd = h_0 ? slice_timestep(h_0, l * nd + 0) : NULL;
-        Tensor** fwd_steps = malloc((size_t)seq_len * sizeof(Tensor*));
+        Tensor** fwd_steps = cml_malloc((size_t)seq_len * sizeof(Tensor*));
```
Comment (jaywyawhare) [inline src/nn/layers/rnn.c:837]: fwd_steps allocation fails, the function returns null without freeing the previously allocated final_h causing a memory leak

File: src/nn/layers/rnn.c [hunk 4]
```diff
@@ -548,14 +549,14 @@ void rnn_forward(RNN* rnn, Tensor* input, Tensor* h_0,
...
-    Tensor** final_h = calloc((size_t)total_dirs, sizeof(Tensor*));
+    Tensor** final_h = cml_calloc((size_t)total_dirs, sizeof(Tensor*));
     Tensor* layer_input = x;
 
     for (int l = 0; l < rnn->num_layers; l++) {
         RNNCell* fwd_cell = rnn->cells[l * nd + 0];
 
         Tensor* h_fwd = h_0 ? slice_timestep(h_0, l * nd + 0) : NULL;
-        Tensor** fwd_steps = malloc((size_t)seq_len * sizeof(Tensor*));
+        Tensor** fwd_steps = cml_malloc((size_t)seq_len * sizeof(Tensor*));
```
Comment (jaywyawhare) [inline src/nn/layers/rnn.c:560]: Same issue from L827

File: src/nn/layers/sequential.c [hunk 5]
```diff
@@ -296,15 +297,15 @@ static void free_cached_graph(CachedModelGraph* cache) {
...
-        free(cache->input_buffer);
+        cml_free(cache->input_buffer);
     }
     if (cache->output_buffer) {
-        free(cache->output_buffer);
+        cml_free(cache->output_buffer);
     }
-    free(cache);
+    cml_free(cache);
 }
```
Comment (jaywyawhare) [inline src/nn/layers/sequential.c:309]: L303 & 306 free cache->input_buffer and cache->output_buffer using cml_free but these buffers are allocated with aligned_alloc at L342 & 351 memory allocated with aligned_alloc must be freed with free(), not a custom allocator

This mismatch can cause heap corruption or crashes

File: src/nn/layers/sequential.c [hunk 6]
```diff
@@ -77,15 +78,15 @@ static void fast_path_free(SequentialFastPath* fp) {
...
+            cml_free(fp->ops[i].out_buf);
         }
     }
-    free(fp->ops);
-    free(fp->input_shape);
-    free(fp);
+    cml_free(fp->ops);
+    cml_free(fp->input_shape);
+    cml_free(fp);
 }
```
Comment (jaywyawhare) [inline src/nn/layers/sequential.c:90]: Same issue from L300 and 308

File: include/alloc/cml_allocator.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+void*  cml_calloc(size_t nmemb, size_t size);
+void*  cml_realloc(void* ptr, size_t new_size);
+void   cml_free(void* ptr);
+
+/* String helper */
+char*  cml_strdup(const char* s);
+
+/* Aligned allocation (alignment must be power of two, >= 16) */
+void*  cml_aligned_alloc(size_t size, size_t alignment);
+void   cml_aligned_free(void* ptr); /* safe to mix with cml_free in most cases */
```
Comment (jaywyawhare) [inline include/alloc/cml_allocator.h]: explain or document when mixing cml_aligned_free with cml_free is unsafe

File: src/backend/blas.c [hunk 8]
```diff
@@ -573,12 +574,12 @@ int cml_blas_sgemm(CMLBlasContext* ctx, const float* A, const float* B, float* C
...
             if (ctx->pack_a_size < need_a) {
-                free(ctx->pack_a_buf);
+                cml_free(ctx->pack_a_buf);
                 ctx->pack_a_buf  = (float*)aligned_alloc(32, need_a);
                 ctx->pack_a_size = ctx->pack_a_buf ? need_a : 0;
             }
             if (ctx->pack_b_size < need_b) {
-                free(ctx->pack_b_buf);
+                cml_free(ctx->pack_b_buf);
                 ctx->pack_b_buf  = (float*)aligned_alloc(32, need_b);
```
Comment (jaywyawhare) [inline src/backend/blas.c]: Same mismatch issue as sequential.c

File: src/datasets/loaders.c [hunk 9]
```diff
@@ -849,25 +850,25 @@ CMLOpenImagesLoader* cml_openimages_open(const char* images_dir, const char* ann
...
+    cml_free(files);
 
-    CMLOpenImagesLoader* loader = calloc(1, sizeof(CMLOpenImagesLoader));
+    CMLOpenImagesLoader* loader = cml_calloc(1, sizeof(CMLOpenImagesLoader));
     loader->image_ids = ids;
     loader->num_images = count;
-    loader->images_dir = strdup(images_dir);
-    loader->annotations_path = strdup(annotations_csv);
+    loader->images_dir = cml_strdup(images_dir);
+    loader->annotations_path = cml_strdup(annotations_csv);
```
Comment (jaywyawhare) [inline src/datasets/loaders.c:882]: Alloc mix up, L862 use std alloc for img id, but later freed using customm cml_free, it can crash with heap curruption

File: src/ops/ir/fused_codegen.c [hunk 10]
```diff
@@ -80,7 +81,7 @@ CMLLinearProgram* cml_linearize_group(const CMLFusionGroup* g) {
                 /* Emit */
                 if (prog->num_ops >= prog->capacity) {
                     int nc = prog->capacity * 2;
-                    CMLLinearOp* tmp = realloc(prog->ops, (size_t)nc * sizeof(CMLLinearOp));
+                    CMLLinearOp* tmp = cml_realloc(prog->ops, (size_t)nc * sizeof(CMLLinearOp));
                     if (!tmp) break;
                     prog->ops = tmp;
                     prog->capacity = nc;
```
Comment (jaywyawhare) [inline src/ops/ir/fused_codegen.c:87]: L85 L117 & L134, allocation failure only breaks but the function still returns prog at L143 that can produce truncated op streams and incorrect fused codegen output

```diff
-                    if (!tmp) break;
+                    if (!tmp) goto oom;

-            if (!tmp) break;
+            if (!tmp) goto oom;

-                if (!tmp) break;
+                if (!tmp) goto oom;

     cml_free(node_vreg);
     return prog;
+
+oom:
+    cml_free(node_vreg);
+    cml_free(prog->ops);
+    cml_free(prog);
+    return NULL;
```

File: src/ops/ir/fusion_schedule.c [hunk 11]
```diff
@@ -324,17 +325,17 @@ CMLFusionSchedule* cml_fusion_schedule_create(CMLGraph_t graph,
...
-            int* dep_counts = calloc((size_t)ng, sizeof(int));
+            int* in_degree = cml_calloc((size_t)ng, sizeof(int));
+            int** deps = cml_calloc((size_t)ng, sizeof(int*));
+            int* dep_counts = cml_calloc((size_t)ng, sizeof(int));
 
             if (in_degree && deps && dep_counts) {
                 for (int i = 0; i < ng; i++) {
-                    deps[i] = calloc((size_t)ng, sizeof(int));
+                    deps[i] = cml_calloc((size_t)ng, sizeof(int));
                 }
```
Comment (jaywyawhare) [inline src/ops/ir/fusion_schedule.c:347]: guard BFS temporary allocations before use and cleanup

if cml_calloc fails here the code still dereferences pointers later L363/374 and also unconditionally iterates deps[i] at L401 this can crash under memory pressure

File: src/ops/ir/graph_cache.c [hunk 12]
```diff
@@ -255,16 +256,16 @@ static void free_execution_plan_impl(CMLExecutionPlan* plan, bool detach_tensor_
                         t->is_executed = false;
                     }
                 }
-                free(plan->buffers[i]);
+                cml_free(plan->buffers[i]);
```
Comment (jaywyawhare) [inline src/ops/ir/graph_cache.c]: plan->buffers[i] is currently allocated with aligned_alloc L228 but freed with cml_free L259

File: tests/test_new_features.c [hunk 13]
```diff
@@ -246,15 +247,15 @@ static int test_muon_optimizer(void) {
...
     if (!param) { tensor_free(weight); return 0; }
     param->tensor = weight;
-    param->name = strdup("test_weight");
+    param->name = cml_strdup("test_weight");
     param->requires_grad = true;
 
     Parameter* params[] = {param};
     Optimizer* opt = optim_muon(params, 1, 0.02f, 0.95f, 0.0f, true);
-    if (!opt) { free(param); tensor_free(weight); return 0; }
+    if (!opt) { cml_free(param); tensor_free(weight); return 0; }
```
Comment (jaywyawhare) [inline tests/test_new_features.c:264]: when optim_muon fails at L257 the cleanup at L258 frees param and weight but does not free param->name this can lead to a memory leak

File: src/nn/layers/activations.c [hunk 14]
```diff
@@ -45,15 +46,15 @@ static Tensor* leaky_relu_forward(Module* module, Tensor* input) {
...
-    LeakyReLU* leaky_relu = malloc(sizeof(LeakyReLU));
+    LeakyReLU* leaky_relu = cml_malloc(sizeof(LeakyReLU));
     if (!leaky_relu)
         return NULL;
 
     if (module_init((Module*)leaky_relu, "LeakyReLU", leaky_relu_forward, leaky_relu_free) != 0) {
-        free(leaky_relu);
+        cml_free(leaky_relu);
         return NULL;
     }
```
Comment (jaywyawhare) [inline src/nn/layers/activations.c:65]: four activation constructors silently return NULL on allocation failure while the other eight activations push diagnostic context to the error stack this can fuck up debugging consider adding error_stack_push calls in these four constructors to match the pattern used by the majority apply similar changes to GELU, Softmax, and LogSoftmax constructors

**Review by jaywyawhare — COMMENTED** [review]
Thanks for contribution, check the reviews and fix the issues
